File: example/customersnode/src/service.ts
import {
  Context,
  Customer,
  CustomerPage,
  CustomerQuery,
  Inbound,
  Outbound,
} from "./interfaces";

export class InboundImpl implements Inbound {
  private outbound: Outbound;

  constructor(outbound: Outbound) {
    this.outbound = outbound;
  }

  async createCustomer(customer: Customer): Promise<Customer> {
    await this.outbound.saveCustomer(customer);
    await this.outbound.customerCreated(customer);

    return customer;
  }

  async getCustomer(id: number): Promise<Customer> {
    const stream = this.outbound.getCustomers();
    await stream.forEach(async (customer) => {
      console.log(customer);
    });
    return this.outbound.fetchCustomer(id);
  }

  async listCustomers(query: CustomerQuery): Promise<CustomerPage> {
    return new CustomerPage({
      offset: query.offset,
      limit: query.limit,
      items: [new Customer()],
    });
  }
}

export class CustomerActorImpl {
  async createCustomer(ctx: Context, customer: Customer): Promise<Customer> {
    ctx.set("customer", customer);
    return customer;
  }

  async getCustomer(ctx: Context): Promise<Customer> {
    const customer = await ctx.get("customer", Customer);
    return customer;
  }
}

File: example/customerspython/service.py
#!/usr/bin/env python3
from adapter import start, register_inbound, register_customer_actor, outbound
from interfaces import Context, Inbound, Customer, CustomerPage, CustomerQuery, CustomerActor


class InboundImpl(Inbound):
    async def create_customer(self, customer: Customer) -> Customer:
        await outbound.save_customer(customer)
        await outbound.customer_created(customer)
        return customer

    async def get_customer(self, id: int) -> Customer:
        return await outbound.fetch_customer(id)

    async def list_customers(self, query: CustomerQuery) -> CustomerPage:
        return CustomerPage(
            offset=query.offset,
            limit=query.limit,
        )


class CustomerActorImpl(CustomerActor):
    async def activate(self, ctx: Context):
        print("ACTIVATED")

    async def create_customer(self, ctx: Context,
                              customer: Customer) -> Customer:
        ctx.set("customer", customer)
        return customer

    async def get_customer(self, ctx: Context) -> Customer:
        return await ctx.get("customer", Customer)

    async def deactivate(self, ctx: Context):
        print("DEACTIVATED")


def main():
    register_inbound(InboundImpl())
    register_customer_actor(CustomerActorImpl())

    start()


if __name__ == "__main__":
    main()
